klippy/extras/ads1x1x.py

```diff
@@ -0,0 +1,431 @@
+        # alert pin is used, which we haven't made configurable.
+        # But that wouldn't be useful for a normal temperature sensor anyway.
+        self.comp_mode = ADS1X1X_COMPARATOR_MODE['TRADITIONAL']
+        self.comp_polarity = ADS1X1X_COMPARATOR_POLARITY['ACTIVE_LO']
+        self.comp_latching = ADS1X1X_COMPARATOR_LATCHING['NON_LATCHING']
+        self.comp_queue = ADS1X1X_COMPARATOR_QUEUE['QUEUE_NONE']
+        self._i2c = bus.MCU_I2C_from_config(config, address)
+
+        self.adc_voltage = config.getfloat('adc_voltage', 5., above=0.)
+
+        self.mcu = self._i2c.get_mcu()
+
+        self._printer.add_object("ads1x1x " + self.name, self)
+        self._printer.register_event_handler("klippy:ready", \
+                                            self._handle_ready)
+        self._printer.register_event_handler("klippy:shutdown", \
+                                            self.reset_all_devices)
+
+        self._pins = {}
+        self._command_queue = []
+
+    def setup_pin(self, pin_type, pin_params):
+        if pin_type == 'adc':
+            pin = ADS1X1X_pin(self, pin_params)
+            if pin.pin in self._pins:
+                raise pins.error(
+                    '%s pin %s for chip %s is used multiple times' \
+                        % (self.chip, pin.pin, self.name))
+            self._pins[pin.pin] = pin
+            return pin
+        raise pins.error('Wrong pin or incompatible type: %s with type %s! ' % (
+            pin_params['pin'], pin_type))
+
+    def _handle_ready(self):
+        self.reset_all_devices()
+        self.sample_timer = \
+                self._reactor.register_timer(self._sample)
+        self._reactor.update_timer(self.sample_timer, self._reactor.NOW)
+
+    def request_sample(self, sensor, callback):
+        if len(self._command_queue) <= ADS1X1_MAX_COMMAND_QUEUE_SIZE:
+            self._command_queue.append({
+                'sensor': sensor,
+                'callback': callback
+            })
+        else:
+            logging.warning(
+                'ADS1X1X: command queue is overflowing, dropping requests')
+
+    def _sample(self, eventtime):
+        if self._command_queue:
+            command = self._command_queue.pop()
+            pin_object = self._pins[command['sensor'].pin]
+            sample = 0
+            try:
+                self._write_register(ADS1X1X_REG_POINTER['CONFIG'],
+                                     pin_object.config)
+
+                self._reactor.pause(self._reactor.monotonic() + 1.0 /
+                                    (self.samples_per_second_numeric + 1))
+
+                sample = self._read_register(ADS1X1X_REG_POINTER['CONVERSION'])
+            except Exception:
+                logging.exception("ADS1X1X: error while sampling")
+                return self._reactor.NEVER
+
+            command['callback'](sample)
+
+        return self._reactor.monotonic() + 1.0 / self.samples_per_second_numeric
```

Comment:
I believe if we're taking samples from multiple ADC pins per sensor then this waketime would cause issues.

Is this specific delay needed if we're pausing above before reading the ADC value?

Reply:
I'll try to check it, but I think you might be right. That looks like an unnecessary double wait.